File: SWEA/D4/1824. 혁진이의 프로그램 검증/혁진이의 프로그램 검증.cpp
#include <iostream>
#include <cstring>
#include <queue>
#include <algorithm>
#include <vector>
#define INF 100000000
#define DOWN 0
#define UP 3
#define LEFT 2
#define RIGHT 1
#define f(a, b, c) for(int a=b;a<c;++a)
using namespace std;
int inp[20][20];
bool v[20][20][16][4]; // R, C, memory_value, direction --> 해당 위치에 memory_value 상태에서 "어느 방향에서 왓는가?"
// ex 오른쪽으로 가서 해당 위치 도착하면 , 도착블럭기준 왼쪽에서 온 것이나 다름없음. 따라서 왼쪽방향에 대한 방문처리함
int mv[4][2] = {{1, 0}, {0, 1}, {0, -1}, {-1, 0}}; //아래쪽, 오른쪽, 왼쪽, 위쪽

struct io {
    int r, c, memo, dir;
};

// int char_command(char ch) {
//     if (ch == '<') return 2; //왼쪽강제이동
//     else if (ch == '>') return 1; //오른쪾강제 이동
//     else if (ch == '^') return 3; //위로
//     else if (ch == 'v') return 0; //아래로

//     else if (ch == '.') return -1; //방향 꺽기 안함.

//     //조건부 방향 꺽기. 0 이냐 아니냐?
//     else if (ch == '_') return 4; // 0 이면 오른쪽, 아님 왼쪽
//     else if (ch == '|') return 5; // 0 이면 아래쪽, 아님 위쪽

//     else if (ch == '?') return 6; // 4방향 전부 다 탐색

//     else if (ch == '@') return -2; // 종료

//     else if (ch == '+') return 7; // memo = (memo + 1)%16
//     else if (ch == '-') return 8; // memo = (memo + 15)%16
    
//     else return ch - '0' + 10; // 숫자입력
// }

int char_command(char ch) {
    switch (ch)
    {
    case '<':
        return 2;
        break;  
    case '>':
        return 1;
        break;
    case '^':
        return 3;
        break;  
    case 'v':
        return 0;
        break;   
    case '.':
        return -1;
        break;   
    case '_':
        return 4;
        break;   
    case '|':
        return 5;
        break;   
    case '?':
        return 6;
        break;   
    case '@':
        return 9;
        break;   
    case '+':
        return 7;
        break;   
    case '-':
        return 8;
        break;   
    default:
        return ch - '0' + 10;
    }
}

int main(void) {
    ios_base::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    // freopen("input.txt","r",stdin);
    int T; cin >> T;
    int R, C;
    string tmp;
    for (int tc = 1; tc <= T; ++tc) {
        cin >> R >> C;
        bool isend = false;
        bool finish = false;
        f(r, 0, R) {
            cin >> tmp;
            f(c, 0, C) {
                inp[r][c] = char_command(tmp[c]);
                if (inp[r][c] == 9) isend = true;
            }   
        }

        //종료 입력 없음 -> 해봤자 의미가 없음
        if (!isend) {
            cout << '#' << tc << " NO\n";
            continue;
        }

        //있으면 시물레이션
        //초기설정
        memset(v, 0, sizeof(v));
        queue<io> q;
        q.push({0, 0, 0, RIGHT}); //가장 좌측 상단. 
        while (!q.empty())
        {   
            int r= q.front().r, c = q.front().c, memo = q.front().memo, dir = q.front().dir;
            q.pop();
            // cout << r <<' ' <<c << ' ' <<memo << ' ' << dir <<'\n';
            
            int command = inp[r][c];

            //종료조건
            if (command == 9) {
                finish = true;
                break;
            }
            else if (command == 6) {
                // cout <<"?\n";
                for (int ndir = 0; ndir < 4; ++ndir) {
                    int nr = (r + mv[ndir][0] + R)%R, nc = (c + mv[ndir][1] + C)%C;
                    // cout << nr << ' ' << nc <<'\n';
                    if (!v[nr][nc][memo][3-ndir]) {
                        v[nr][nc][memo][3-ndir] = true;
                        q.push({nr, nc, memo, ndir});
                    }
                }
            }
            else {
                if (command <= 3 ) {
                    if (command >= 0) dir = command;
                }
                else if (command == 4) dir = (memo == 0 ? RIGHT : LEFT);
                else if (command == 5) dir = (memo == 0 ? DOWN : UP);
                else if (command == 7) memo = (memo + 1)%16;
                else if (command == 8) memo = (memo + 15)%16;
                else memo = command - 10;

                int nr = (r + mv[dir][0] + R)%R, nc = (c + mv[dir][1] + C)%C;
                if (!v[nr][nc][memo][3-dir]) {
                    v[nr][nc][memo][3-dir] = true;
                    q.push({nr, nc, memo, dir});
                }                
            }
        }

        if (finish) cout << '#' << tc << " YES\n";
        else cout << '#' << tc << " NO\n";
    }
    return 0;
}

File: SWEA/D4/1824. 혁진이의 프로그램 검증/혁진이의 프로그램 검증.py
from collections import deque

#방향상수
DOWN = 0
RIGHT = 1
LEFT = 2
UP = 3

mv = [(1, 0), (0, 1), (0, -1), (-1, 0)] # DOWN, RIGHT, LEFT, UP
# p.s. 3 - 방향 = 반대방향

def char_cmd(ch : str) -> int:
    if ch == '<': return LEFT
    elif ch == '>': return RIGHT
    elif ch == '^': return UP
    elif ch == 'v': return DOWN
    elif ch == '.': return -1
    elif ch == '_': return 4
    elif ch == '|': return 5
    elif ch == '?': return 6
    elif ch == '@': return 9
    elif ch == '+': return 7
    elif ch == '-': return 8
    else: return ord(ch) - ord('0') + 10


T = int(input())
for tc in range(1, T+1):
    R, C = map(int,input().split())
    cmd_map = [ input().rstrip() for _ in range(R) ]
    
    isend= False
    for inp in cmd_map:
        if '@' in inp:
            isend = True
            break
    if not isend: #종료 입력 자체가 없으면 어차피 절대 안끝남 -> 조기종료
        print(f'#{tc} NO')
        continue
    
    # v[행][열][메모리값][방향], 방향은 여기서 해당 블럭에 들어오는 방향으로 잡음
    v = [ [ [ [False] * 4 for _ in range(16) ] for _ in range(C) ] for _ in range(R) ]
    dq = deque()
    dq.append(tuple([0, 0, 0, RIGHT]))
    finish = False #도착했다는 의미
    
    #먼저 해당 명령어를 실행하고, 다음 방향으로 이동하는 방식
    while dq:
        r, c, memo, dir = dq.popleft()
        cmd = char_cmd(cmd_map[r][c])
        
        if cmd == 9: #도착도달. 종료
            finish = True
            break
        
        elif cmd == 6:  # ?처리
            for idx in range(4):
                #다음이동, 맵 범위 넘어가면 반대편으로 넘어가는것 고려
                nr = (r + mv[idx][0] + R)%R
                nc = (c + mv[idx][1] + C)%C
                
                if not v[nr][nc][memo][3-idx]:
                    v[nr][nc][memo][3-idx] = True
                    dq.append((nr, nc, memo, idx))
        else:
            #그 외 명령어들. 단 1개의 dq push처리되는 것
            
            #방향돌리기
            if cmd <= 3:
                if cmd >= 0: dir = cmd

            elif cmd == 4: #_
                dir = RIGHT if memo == 0 else LEFT
            elif cmd == 5: #|
                dir = DOWN if memo == 0 else UP
            elif cmd == 7: #+
                memo = (memo + 1)%16
            elif cmd == 8: #-
                memo = (memo + 15) % 16
            else: #수 입력
                memo = cmd - 10
            
            nr = (r + mv[dir][0] + R)%R
            nc = (c + mv[dir][1] + C)%C
            
            if not v[nr][nc][memo][3-dir]:
                v[nr][nc][memo][3-dir] = True
                dq.append((nr, nc, memo, dir))
    
    print(f'#{tc} {"YES" if finish else "NO"}')